chain/xrp/client/client.go

```diff
@@ -274,6 +411,147 @@ type NewFields struct {
+	}
+
+	if mnw.node.LedgerEntryType == "AccountRoot" {
+		return xc.Address(mnw.node.FinalFields.Account), nil
+
+	} else if mnw.node.LedgerEntryType == "RippleState" {
+		isSource, fetchIsSourceErr := mnw.IsSource(txResponse)
+		if fetchIsSourceErr != nil {
+			return "", fetchIsSourceErr
+		}
+
+		if isSource {
+			if mnw.node.FinalFields.LowLimit == nil {
+				return "", fmt.Errorf("empty HighLimit in FinalFields")
+			}
+
+			return xc.Address(mnw.node.FinalFields.LowLimit.Issuer), nil
+		} else {
+			if mnw.node.FinalFields.HighLimit == nil {
+				return "", fmt.Errorf("empty HighLimit in FinalFields")
+			}
+
+			return xc.Address(mnw.node.FinalFields.HighLimit.Issuer), nil
+		}
+	}
+
+	return "", fmt.Errorf("unknown node type in ModifiedNode")
+}
+
+func (mnw *ModifiedNodeWrapper) GetContract(txResponse *TransactionResponse) (xc.ContractAddress, error) {
+	var (
+		contract                  xc.ContractAddress
+		finalBalanceHumanReadable xc.AmountHumanReadable
+		err                       error
+	)
+
+	if mnw.node.FinalFields == nil {
+		return "", fmt.Errorf("empty FinalFields in ModifiedNode")
+	}
+
+	if mnw.node.FinalFields.Balance.XRPAmount != "" {
+		contract = ""
+	} else {
+		finalBalanceHumanReadable, err = xc.NewAmountHumanReadableFromStr(mnw.node.FinalFields.Balance.TokenAmount.Value)
+		if err != nil {
+			return "", err
+		}
+
+		finalBalanceBlockchain := finalBalanceHumanReadable.ToBlockchain(6)
+		zero := xc.NewAmountBlockchainFromUint64(0)
+
+		if finalBalanceBlockchain.Cmp(&zero) > 0 {
+			if mnw.node.FinalFields.HighLimit == nil {
+				return "", fmt.Errorf("empty HighLimit in FinalFields")
+			}
+
+			contract = xc.ContractAddress(mnw.node.FinalFields.HighLimit.Issuer)
+		} else {
+			if mnw.node.FinalFields.LowLimit == nil {
+				return "", fmt.Errorf("empty LowLimit in FinalFields")
+			}
+
+			contract = xc.ContractAddress(mnw.node.FinalFields.LowLimit.Issuer)
+		}
+	}
+
+	return contract, nil
+}
+
+func (mnw *ModifiedNodeWrapper) GetAmount() (xc.AmountBlockchain, error) {
+	transactedAmount, conversionErr := ExtractModifiedNodeBalance(mnw.node)
+	if conversionErr != nil {
+		return xc.AmountBlockchain{}, fmt.Errorf("failed to fetch ModifiedNode balance: %s", conversionErr.Error())
+	}
+
+	return transactedAmount, nil
+}
+
+func (mnw *ModifiedNodeWrapper) IsSource(txResponse *TransactionResponse) (bool, error) {
+	if mnw.node.LedgerEntryType == "AccountRoot" {
+		if mnw.node.FinalFields == nil {
+			return false, fmt.Errorf("empty FinalField in ModifiedNode")
+		}
+
+		if mnw.node.FinalFields.Account != txResponse.Result.Account {
+			return false, nil
+		} else {
+			return true, nil
+		}
+
+	} else if mnw.node.LedgerEntryType == "RippleState" {
+		finalBalanceHumanReadable, err := xc.NewAmountHumanReadableFromStr(mnw.node.FinalFields.Balance.TokenAmount.Value)
+		if err != nil {
+			return false, err
+		}
+
+		finalBalanceBlockchain := finalBalanceHumanReadable.ToBlockchain(6)
```

Comment:
may not matter since comparing to zero, but this is the wrong decimals i believe.

Reply:
we should use the max precision (15).  because if it's a token then there's a good chance it will get truncated.

E.g. if the amount is `0.0000000000001554354`, and we use this with 6 decimals, then we will get `0` out.

Reply:
The underlying type for `xc.AmountHuman` is `decimal.Decimal` which has more methods we can use.  Perhaps let's just do this.

```go
isZero := decimal.Decimal(finalBalanceHumanReadable).IsZero()
```